source/fuzz/transformation_duplicate_region_with_selection.cpp

```diff
@@ -0,0 +1,566 @@
+  // transformed region will end in |merge_block|. Hence, we need to change all
+  // occurrences of the label id of the |exit_block| to the label id of the
+  // |merge_block|.
+  exit_block->ForEachSuccessorLabel([this, ir_context](uint32_t label_id) {
+    auto block = ir_context->cfg()->block(label_id);
+    for (auto& instr : *block) {
+      if (instr.opcode() == SpvOpPhi) {
+        instr.ForEachId([this](uint32_t* id) {
+          if (*id == message_.exit_block_id()) {
+            *id = message_.merge_label_fresh_id();
+          }
+        });
+      }
+    }
+  });
+
+  // Similarly, we need to update OpPhi instructions in the |entry_block|. We
+  // know that it has only one predecessor, since the region is single-entry,
+  // single-exit. Its constructs and their merge blocks must be either wholly
+  // within or wholly outside of the region. We need to change all occurrences
+  // of its id to the label id of the |new_entry| block.
+  uint32_t entry_block_pred_id =
+      ir_context
+          ->get_instr_block(ir_context->cfg()->preds(entry_block->id())[0])
+          ->id();
+  for (auto& instr : *entry_block) {
+    if (instr.opcode() == SpvOpPhi) {
+      instr.ForEachId([this, entry_block_pred_id](uint32_t* id) {
+        if (*id == entry_block_pred_id) {
+          *id = message_.new_entry_fresh_id();
+        }
+      });
+    }
+  }
+
+  // Duplication of blocks will invalidate iterators. Store all the blocks from
+  // the enclosing function.
+  std::vector<opt::BasicBlock*> blocks;
+  for (auto& block : *enclosing_function) {
+    blocks.push_back(&block);
+  }
+  // Store instructions OpReturn, OpReturnValue, OpBranch from the last block to
+  // be moved (cloned and removed).
+  std::vector<opt::Instruction*> instructions_to_move;
+  opt::BasicBlock* previous_block = nullptr;
+
+  // Iterate over all blocks of the function to duplicate blocks of the original
+  // region and their instructions.
+  for (auto& block : blocks) {
+    // The block must be contained in the region.
+    if (region_blocks.count(block) == 0) {
+      continue;
+    }
+
+    fuzzerutil::UpdateModuleIdBound(
+        ir_context, original_label_to_duplicate_label[block->id()]);
+
+    std::unique_ptr<opt::BasicBlock> duplicated_block =
+        MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(
+            ir_context, SpvOpLabel, 0,
+            original_label_to_duplicate_label[block->id()],
+            opt::Instruction::OperandList()));
+
+    for (auto& instr : *block) {
+      // Cases where an instruction is an OpBranch
+      // or an OpReturn or an OpReturnValue in the exit block are handled
+      // separately, after the iteration.
+      if (block == exit_block && (instr.opcode() == SpvOpReturn ||
+                                  instr.opcode() == SpvOpReturnValue ||
+                                  instr.opcode() == SpvOpBranch)) {
+        instructions_to_move.push_back(&instr);
+        continue;
```

Comment:
A couple of thoughts here:

(1) Are these the only block terminators the exit block can have?  Could it have another termination instruction such as OpBranchConditional, OpKill, OpSwitch, OpUnreachable?

I think it would be good to check whether the instructions opcode is a block terminator, and if it is, assert that it is one of the ones that you expect.

(2) Could you get rid of `instructions_to_move` and simply `continue` at this point?  Then, where you currently use `instructions_to_move`, you could look directly at the terminator of `exit_block` - it has exactly the instruction you need to move.

I think it's misleading to have `instructions_to_move` when - if I have understood correctly - there will only ever be *one* instruction to move.

Reply:
Can the exit block end with OpBranchConditional, OpKill, OpSwitch, OpUnreacheable? I am not sure. If so, what should I assert?

Reply:
I think that it cannot end with neither OpBranchConditional nor OpSwitch, since the region is single-entry, single-exit. But I'm not sure about the other two cases.

Reply:
In general, when writing code to handle a situation where data values come from a set S, but you know (for whatever reason) that they will only have values from some subset T of S, it's good to (1) assert that the data value is in T, then (2) handle just the cases from T.

So here, if you think SpvOpReturn, SpvOpReturnValue and SpvOpBranch are the only allowed values of the terminator of exit block, you should do something like:

```
if (block == exit_block && spvOpcodeIsBlockTerminator(instr.opcode())) {
  switch(instr.opcode()) {
     case SpvOpBranch:
     case SpvOpReturn:
     case SpvOpReturnValue:
       // Code to handle the expected case
       break;
     default:
       assert(false && "Unexpected terminator for exit block of region.")
       break;
  }
}
```
  
Now, the question is: are `OpUnreachable` and `OpKill` allowed?

To answer that, write an example for each case, where the region you're trying to duplicate ends with the instruction.

I think it *should* be possible to duplicate a region that ends with one of these; I can't really see any reason why not.  Either way, it would be good to have a test that shows it is applicable and that it works, or a test that confirms it's not applicable.

Reply:
Thanks for your explanation. Tests have confirmed that OpUnreachable and OpKill are allowed.